src/renderer.ts

```diff
@@ -68,70 +77,101 @@ export class Renderer {
+            clef = this.clefTranslate(value);
             t.literal(`clef = '${clef}'`);
-            if (clef !== '') {
-              t.expression(() => stave!.addClef(clef));
+            if (duration == 0) {
+              t.literal(`staff = ${staff}`);
+              t.expression(() => staves.get(staff)!.addClef(clef));
+            } else {
+              t.expression(() => notes.push(factory.ClefNote({ type: clef, options: { size: 'small' } })));
             }
-            if (timeSignature !== '/') {
-              t.literal(`timeSignature = '${timeSignature}'`);
-              t.expression(() => stave!.addTimeSignature(timeSignature));
-            }
-          }
-          t.expression(() => (voices = []));
+          });
+
+          timeSignature = message.time!;
+          t.literal(`timeSignature = '${timeSignature}'`);
+          if (timeSignature)
+            t.expression(() =>
+              staves.forEach((stave) => {
+                stave.addTimeSignature(timeSignature);
+              })
+            );
+          break;
+        case 'voiceEnd':
+          voice = parseInt(message.voice);
+          t.literal(`voice = ${voice}`);
+          t.expression(() => voices.set(voice, factory.Voice().setMode(2).addTickables(notes)));
+          t.expression(() => (notes = []));
+          duration = 0;
           break;
         case 'note':
           const durationDenominator = this.getDurationDenominator(message.type);
-          let name = '';
-          let options = {};
+          const noteStruct: VF.GraceNoteStruct = {};
 
-          options = { stem: message.stem, clef: this.getClef(curClefs, message.staff, 'treble') };
+          if (message.stem) noteStruct.stem_direction = message.stem == 'up' ? 1 : -1;
+          noteStruct.clef = this.clefGet(message.staff, duration);
+          if (message.duration) duration += message.duration;
           // no pitch, rest
           if (message.head.length == 0) {
-            name = `B4/${durationDenominator}/r`;
-            options = {};
-            // single pitch
-          } else if (message.head.length == 1) name = `${message.head[0].pitch}/${durationDenominator}`;
-          // multiple pitches, chords
-          else if (message.head.length > 1) {
+            if (noteStruct.clef == 'bass') noteStruct.keys = ['D/2'];
+            else noteStruct.keys = ['B/4'];
+            noteStruct.duration = `${durationDenominator}r`;
+          } else {
+            noteStruct.keys = [];
             for (const head of message.head) {
-              name += ` ${head.pitch}`;
+              noteStruct.keys.push(`${head.pitch}`);
             }
-            name = name.replace(/ /, '(');
-            name += `)/${durationDenominator}`;
+            noteStruct.duration = `${durationDenominator}`;
           }
 
+          if (message.grace) {
+            noteStruct.slash = message.graceSlash;
+            note = this.factory.GraceNote(noteStruct);
```

Comment:
nit: To avoid typing it twice here and the instance below, I think this could be

```ts
t.expression(() => note = factory.GraceNote(noteStruct));
```

Reply:
You would need to define notestruct then. I will leave it as it. I think we have to look at the output code anyway, it is getting too complex.